File: Java/Medium/Accepted/time-needed-to-inform-all-employees_0.java
class Solution {
    public int numOfMinutes(int n, int headID, int[] manager, int[] informTime) {
        
        Map<Integer, List<Integer>> map = new HashMap<>();
        for (int i = 0; i < manager.length; i++) {
            if (!map.containsKey(manager[i])) {
                map.put(manager[i], new ArrayList<>());
            }
            map.get(manager[i]).add(i);
        }
        return dfs(map, headID, informTime);
    }
    
    private int dfs(Map<Integer, List<Integer>> map, int id, int[] informTime) {
        if (!map.containsKey(id)) {
            return 0;
        }
        int max = 0;
        for (int sub : map.get(id)) {
            max = Math.max(max, dfs(map, sub, informTime));
        }
        return max + informTime[id];
    }
}

File: C++/Medium/Accepted/time-needed-to-inform-all-employees_0.cpp
class Solution {
public:
    int numOfMinutes(int n, int headID, vector<int>& manager, vector<int>& informTime) {
        
        vector<vector<int>> tree(n);
        for (int i = 0; i < n; ++i) {
            if (manager[i] != -1) {
                tree[manager[i]].push_back(i);
            }
        }
        return dfs(tree, informTime, headID);
    }
    
    int dfs(vector<vector<int>>& tree, vector<int>& informTime, int root) {
        int res = 0;
        for (int child: tree[root]) {
            res = max(res, dfs(tree, informTime, child));
        }
        return res + informTime[root];
    }
};